arraycontext/impl/cupy/__init__.py

```diff
@@ -0,0 +1,207 @@
+    Array,
+    ArrayContext,
+    ArrayOrContainerOrScalar,
+    ArrayOrContainerOrScalarT,
+    ContainerOrScalarT,
+    NumpyOrContainerOrScalar,
+    UntransformedCodeWarning,
+)
+
+
+class CupyNonObjectArrayMetaclass(type):
+    def __instancecheck__(cls, instance: Any) -> bool:
+        import cupy as cp  # type: ignore[import-untyped]
+        return isinstance(instance, cp.ndarray) and instance.dtype != object
+
+
+class CupyNonObjectArray(metaclass=CupyNonObjectArrayMetaclass):
+    pass
+
+
+class CupyArrayContext(ArrayContext):
+    """
+    An :class:`ArrayContext` that uses :class:`cupy.ndarray` to represent arrays.
+
+    .. automethod:: __init__
+    """
+
+    _loopy_transform_cache: dict[lp.TranslationUnit, lp.ExecutorBase]
+
+    def __init__(self, device: int | None = None) -> None:
+        super().__init__()
+        self._loopy_transform_cache = {}
+
+        if device is not None:
+            import cupy as cp
+            cp.cuda.runtime.setDevice(device)
+
+    array_types = (CupyNonObjectArray,)
+
+    def _get_fake_numpy_namespace(self):
+        from .fake_numpy import CupyFakeNumpyNamespace
+        return CupyFakeNumpyNamespace(self)
+
+    # {{{ ArrayContext interface
+
+    def clone(self):
+        return type(self)()
+
+    @overload
+    def from_numpy(self, array: np.ndarray) -> Array:
+        ...
+
+    @overload
+    def from_numpy(self, array: ContainerOrScalarT) -> ContainerOrScalarT:
+        ...
+
+    def from_numpy(self,
+                   array: NumpyOrContainerOrScalar
+                   ) -> ArrayOrContainerOrScalar:
+        import cupy as cp
+
+        def _from_numpy(ary):
+            return cp.array(ary)
+
+        return with_array_context(rec_map_array_container(_from_numpy, array),
+                                  actx=self)
+
+    @overload
+    def to_numpy(self, array: Array) -> np.ndarray:
+        ...
+
+    @overload
+    def to_numpy(self, array: ContainerOrScalarT) -> ContainerOrScalarT:
+        ...
+
+    def to_numpy(self,
+                 array: ArrayOrContainerOrScalar
+                 ) -> NumpyOrContainerOrScalar:
+        import cupy as cp
+
+        def _to_numpy(ary):
+            return cp.asnumpy(ary)
+
+        return with_array_context(rec_map_array_container(_to_numpy, array),
+                                  actx=None)
+
+    def call_loopy(
+                self,
+                t_unit: lp.TranslationUnit, **kwargs: Any
+            ) -> dict[str, Array]:
+        t_unit = t_unit.copy(target=lp.ExecutableCTarget())
+        try:
+            executor = self._loopy_transform_cache[t_unit]
+        except KeyError:
+            executor = self.transform_loopy_program(t_unit).executor()
+            self._loopy_transform_cache[t_unit] = executor
+
+        _, result = executor(**kwargs)
+
+        return result
+
+    def freeze(self, array):
+        import cupy as cp
+
+        def _freeze(ary):
+            return cp.asnumpy(ary)
```

Comment:
Does `cupy` have something like `jax.block_until_ready` or `pyopencl.finish`? This transfers the memory back to the host, right?

Reply:
it does - we could synchronize on the (default) stream, for example. Not sure if it's worth the extra effort though.

Reply:
I was mostly worried that copying the memory from the device and back on freeze+thaw could hurt performance? Is that not an issue?

Might be worth adding a comment here to explain why this choice was made.

Reply:
I did try running mirgecom with a nonblocking `cp.asnumpy` plus a stream synchronization, the time step time was not affected at all. I added a brief comment in 69e2133